# Import Libraries

In [1]:
import numpy as np

In [2]:
X1 = np.array([1, 2, 3, 4, 5, 6])
Y = np.array([3000, 3250, 3500, 3750, 4000, 4250])
m= len(X1)

print(f'Years of work: {X1}')
print(f'Salary: {Y}')
print(f'Number of Samples: {m}')

Years of work: [1 2 3 4 5 6]
Salary: [3000 3250 3500 3750 4000 4250]
Number of Samples: 6


In [4]:
X1=X1.reshape(m, 1)
print(X1)
print(X1.shape)

[[1]
 [2]
 [3]
 [4]
 [5]
 [6]]
(6, 1)


In [5]:
bias = np.ones((m,1))
print(bias)
print(bias.shape)

[[1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]]
(6, 1)


In [6]:
X = np.append(bias,X1,axis=1)
print(X)
print(X.shape)

[[1. 1.]
 [1. 2.]
 [1. 3.]
 [1. 4.]
 [1. 5.]
 [1. 6.]]
(6, 2)


# Normal Equation

linear regression in $R^2$:
$$ Y=w_0 +w_1X_1 $$
$$ Y=W^TX $$
where:
$$ \mathbf{W} =
\left[ \begin{array}{c}
w_{0}  \\
w_{1} 
\end{array} \right] $$

$$\mathbf{X} =
\left[ \begin{array}{c}
1  \\
X_{1} 
\end{array} \right] $$

thereform:
$$ W^T=[w_0 \text{ } w_1]  $$

$$ Y=W^T X=[w_0 \text{ }w_1] \cdot \mathbf{X} =
\left[ \begin{array}{c}
1  \\
X_{1} 
\end{array} \right] =w_0+w_1X1 $$
Normal Equation - equation for calculating the minimun loss function
$$ W=(X^T X)^{-1}(X^TY)$$

In [7]:
np.dot(X.T, X)

array([[ 6., 21.],
       [21., 91.]])

In [8]:
L= np.linalg.inv(np.dot(X.T, X))
L

array([[ 0.86666667, -0.2       ],
       [-0.2       ,  0.05714286]])

In [9]:
R = np.dot(X.T, Y)
R

array([21750., 80500.])

In [10]:
np.dot(L, R)

array([2750.,  250.])

Solution of the equation
$$ Y=2750+250X_1 $$

# Linear regression using a library scikit-learn

In [11]:
from sklearn.linear_model import LinearRegression

regression = LinearRegression()
regression.fit(X1, Y)

print(regression.intercept_)
print(regression.coef_[0])

2750.0
250.0
